editor/src/clj/editor/asset_browser.clj

```diff
@@ -357,32 +358,55 @@
-
 (defn rename? [resources]
   (and (disk-availability/available?)
-       (= 1 (count resources))
-       (let [resource (first resources)]
-         (and (resource/editable? resource)
-              (not (resource/read-only? resource))
-              (not (fixed-resource-paths (resource/resource->proj-path resource)))))))
+       (pos? (count resources))
+       (every? (fn [resource]
+                 (and (resource/editable? resource)
+                      (not (resource/read-only? resource))
+                      (not (fixed-resource-paths (resource/resource->proj-path resource)))))
+               resources)
+       (case (into #{} (map resource/source-type) resources)
+         #{:folder} (= 1 (count resources))
+         #{:file} (and
+                    (= 1 (count (into #{} (map resource/base-name) resources)))
+                    (= (count resources) (count (into #{} (map resource/ext) resources))))
+         false)))
+
+(defn rename [resources new-name]
+  {:pre [(string? new-name) (rename? resources)]}
+  (let [workspace (resource/workspace (first resources))
+        project-directory-file (workspace/project-path workspace)
+        dir (= :folder (resource/source-type (first resources)))
+        rename-pairs (mapv
+                       (fn [resource]
+                         (let [resource-file (io/file resource)
+                               parent (.getParent resource-file)
+                               ext (resource/ext resource)]
+                           (pair resource-file
+                                 (io/file parent (cond-> new-name
+                                                         (and (not dir) (seq ext))
+                                                         (str "." ext))))))
+                       resources)]
+    (when-not (some #(resource-watch/reserved-proj-path?
+                       project-directory-file
+                       (resource/file->proj-path project-directory-file (val %)))
+                    rename-pairs)
+      ;; plain case change causes irrelevant conflict on case-insensitive file systems
+      ;; fs/move! handles this, no need to resolve
+      (let [{case-changes true possible-conflicts false}
+            (group-by #(fs/same-file? (key %) (val %)) rename-pairs)]
+        (when-let [resolved-conflicts (resolve-any-conflicts possible-conflicts)]
+          (let [resolved-rename-pairs (into resolved-conflicts case-changes)]
+            (when (seq resolved-rename-pairs)
+              (workspace/resource-sync!
+                workspace
+                (into []
+                      (mapcat (fn [[src-file dst-file]]
+                                (let [src-files (vec (file-seq src-file))]
+                                  (fs/move! src-file dst-file)
+                                  (moved-files src-file dst-file src-files))))
+                      resolved-rename-pairs)))))))))
```

Comment:
I moved `rename?` and `rename` around so I can assert that `rename?` holds when `rename` is called.

Reply:
The `rename` logic was changed to allow renaming multiple files at once, but the intent and behavior largely stayed the same.